arch/riscv/core/cache.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright 2023 Andestech
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/cache.h>
+#include <zephyr/sys/barrier.h>
+
+static struct k_spinlock lock;
+
+#ifdef CONFIG_DCACHE
+
+extern size_t __soc_get_dcache_line_size(void);
+extern void __soc_dcache_enable(void);
+extern void __soc_dcache_disable(void);
+size_t __weak __soc_get_dcache_line_size(void) {}
+void __weak __soc_dcache_enable(void) {}
+void __weak __soc_dcache_disable(void) {}
```

Comment:
The usage of the `CONFIG_*` symbols in this file is all wrong, please check https://docs.zephyrproject.org/latest/hardware/cache/index.html about how to use the `CONFIG_*` symbols related to cache.

If `!CONFIG_DCACHE` that means that cache is not supported at all so all the functions must return `-ENOTSUP`. Here you are calling some `weak` or `external` functions so this should be dependent on `CONFIG_EXTERNAL_CACHE`.

Also using non-inlined `.c` file to implement cache functions is at least discouraged. Please check how this is done for example in `include/zephyr/arch/arm64/cache.h`

Reply:
Thanks for your replying.
We think the weak or external is not for CONFIG_EXTERNAL_CACHE, because the riscv CMO extension is only supporting the cache operation(writeback, invalid and flush).
So we think that there may be neccessary for use the soc hook function, and the hook function is provided by SoC, not from the driver cache.
I would like to know if our understanding is not correctly.

Reply:
> We think the weak or external is not for CONFIG_EXTERNAL_CACHE, because the riscv CMO extension is only supporting the cache operation(writeback, invalid and flush).

Then just return `-ENOTSUP` when not supported.

The problem here is that you are writing a cache driver for ARCH cache operations (see `CONFIG_ARCH_CACHE`) but using some SOC-specific external functions and that makes no sense.

Reply:
After the discussion, we think following the rule that doesn't use hook function in cache operation is a good direction, so we commit a new version for that.
But we think that the hook function may still be needed in the step of cache_init  to allow the SOC to initialize their caches.
Also we implement the api as inline function like include/zephyr/arch/arm64/cache.h to replace the previous version.